Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -0,0 +1,498 @@
+    /// used by the `connectionProviders` dictionary to allow retreiving and creating
+    /// connection providers associated to a certain request in constant time.
+    struct Key: Hashable {
+        init(_ request: HTTPClient.Request) {
+            switch request.scheme {
+            case "http":
+                self.scheme = .http
+            case "https":
+                self.scheme = .https
+            default:
+                fatalError("HTTPClient.Request scheme should already be a valid one")
+            }
+            self.port = request.port
+            self.host = request.host
+        }
+
+        var scheme: Scheme
+        var host: String
+        var port: Int
+
+        enum Scheme: Hashable {
+            case http
+            case https
+        }
+    }
+
+    /// A `Connection` represents a `Channel` in the context of the connection pool
+    ///
+    /// In the `ConnectionPool`, each `Channel` belongs to a given `HTTP1ConnectionProvider`
+    /// and has a certain "lease state" (see the `isLeased` property).
+    /// The role of `Connection` is to model this by storing a `Channel` alongside its associated properties
+    /// so that they can be passed around together.
+    class Connection {
+        init(key: Key, channel: Channel, parentPool: ConnectionPool) {
+            self.key = key
+            self.channel = channel
+            self.parentPool = parentPool
+        }
+
+        /// Release this `Connection` to its associated `HTTP1ConnectionProvider` in the parent `ConnectionPool`
+        ///
+        /// This is exactly equivalent to calling `.release(theProvider)` on `ConnectionPool`
+        ///
+        /// - Warning: This only releases the connection and doesn't take care of cleaning handlers in th
+        ///  `Channel` pipeline.
+        func release() {
+            self.parentPool.release(self)
+        }
+
+        /// The connection pool this `Connection` belongs to.
+        ///
+        /// This enables calling methods like `release()` directly on a `Connection` instead of
+        /// calling `pool.release(connection)`. This gives a more object oriented feel to the API
+        /// and can avoid having to keep explicit references to the pool at call site.
+        let parentPool: ConnectionPool
+
+        /// The `Key` of the `HTTP1ConnectionProvider` this `Connection` belongs to
+        ///
+        /// This lets `ConnectionPool` know the relationship between `Connection`s and `HTTP1ConnectionProvider`s
+        fileprivate let key: Key
+
+        /// The `Channel` of this `Connection`
+        ///
+        /// - Warning: Requests that lease connections from the `ConnectionPool` are responsible
+        /// for removing the specific handlers they added to the `Channel` pipeline before releasing it to the pool.
+        let channel: Channel
+
+        /// Wether the connection is currently leased or not
+        var isLeased: NIOAtomic<Bool> = .makeAtomic(value: false)
```

Comment:
I don't think this needs to be an atomic, as far as I can tell this is only ever accessed under a lock. If we do ever access it not under the right lock, it's a race anyway.

I would just document that `Connection` is not thread-safe and that its caller is responsible to only ever access it under the right lock.

Reply:
https://github.com/swift-server/async-http-client/pull/105/commits/9cba7c4439ff4aad2dae4f627eb73691190f0343 should solve this, thanks!